### DISCUSSION TITLE
Comparison: Inpaint with Xinsir Union ControlNet

### DISCUSSION THREAD

**Discussion by Acly** [Acly/krita-ai-diffusion#950] [Research]
I'm testing the inpaint mode of the latest ["Union" ControlNet by Xinsir](https://huggingface.co/xinsir/controlnet-union-sdxl-1.0). This is a comparison to the Fooocus inpaint patch used at the moment (which I believe is based on [Diffusers Inpainting model](https://huggingface.co/diffusers/stable-diffusion-xl-1.0-inpainting-0.1)). 

## Test Setup

All tests use JuggernautXL. The checkpoint is not important here, both methods work with all SDXL checkpoints (excluding Pony derivatives). Tests also use the same default settings (DPM++ 2M Karras at 20 samples), same pre-processing, and identical seeds. 

ControlNet strength is at 0.5 for the results below (the only case where 1.0 was better is the bear that was cut off).

## Apple tree - illustration - outpaint (expand)
Left: IP-Adapter | Right: "children's illustration of kids next to an apple tree"

![benchmark_inpaint_apple-tree_input](https://github.com/user-attachments/assets/d6298850-efac-4171-b4e1-31d67741c5cc)

Xinsir Union ControlNet
![benchmark_inpaint_apple-tree](https://github.com/user-attachments/assets/e4f755aa-994b-4fed-9f15-2db7b02105c0)

Diffusers Inpainting (Fooocus patch)
![benchmark_inpaint_apple-tree](https://github.com/user-attachments/assets/176472c5-40f5-40fc-bb49-ae6b3441d607)

## Canal - photo - inpaint (fill)
Left: IP-Adapter | Right: "photo of a canal in bruges, belgium"

![benchmark_inpaint_bruges_input](https://github.com/user-attachments/assets/2e8ae4ab-cddc-4250-969d-a5646dd7544e)

Xinsir Union ControlNet
![benchmark_inpaint_bruges](https://github.com/user-attachments/assets/c80c3a32-851e-4ce5-b928-078eb1090623)

Diffusers Inpainting (Fooocus patch)
![benchmark_inpaint_bruges](https://github.com/user-attachments/assets/d2304245-30af-4d88-84bb-f5153cc30f85)

## Cornfield - anime - outpaint (expand)
Left: IP-Adapter | Right: "anime artwork of girl in a cornfield"

![benchmark_inpaint_girl-cornfield_input](https://github.com/user-attachments/assets/769a46d5-a55e-481d-96fe-ce08bae37aeb)

Xinsir Union ControlNet
![benchmark_inpaint_girl-cornfield](https://github.com/user-attachments/assets/71a32c5c-0be6-43bd-9ac3-8e680e14aead)

Diffusers Inpainting (Fooocus patch)
![benchmark_inpaint_girl-cornfield](https://github.com/user-attachments/assets/a22a16fb-a48d-44fd-977d-08b5eda7c3e3)

## Jungle - painting - inpaint (fill)
Left: IP-Adapter | Right: "concept artwork of a lake in a forest"

![benchmark_inpaint_jungle_input](https://github.com/user-attachments/assets/4f9e3a5d-f349-4e38-907a-7d253cdbfce7)

Xinsir Union ControlNet
![benchmark_inpaint_jungle](https://github.com/user-attachments/assets/26880fc1-5f23-47b5-9723-d67a6ee3db97)

Diffusers Inpainting (Fooocus patch)
![benchmark_inpaint_jungle](https://github.com/user-attachments/assets/acdb0b9e-5fb7-4a53-9efe-bf1e552289f2)

## Nature - photo - inpaint (add object)
"photo of a black bear standing in a stony river bed"

![benchmark_inpaint_nature_input](https://github.com/user-attachments/assets/df14d037-0aa9-40d2-b49d-c6cc974f0731)

Xinsir Union ControlNet
![benchmark_inpaint_nature](https://github.com/user-attachments/assets/a51246c4-5fb1-4e2e-b121-19f4c726ed70)

Diffusers Inpainting (Fooocus patch)
![benchmark_inpaint_nature](https://github.com/user-attachments/assets/9e04bf3b-22b2-47d1-8755-8c96c157421b)

## Park - photo - inpaint (add object)
"photo of a lady sitting on a bench in a park"

![benchmark_inpaint_park_input](https://github.com/user-attachments/assets/ae11cceb-6f7c-493f-88b9-de8891c023e4)

Xinsir Union ControlNet
![benchmark_inpaint_park](https://github.com/user-attachments/assets/4f056134-d9bc-4948-9a28-d067c834ad12)

Diffusers Inpainting (Fooocus patch)
![benchmark_inpaint_park](https://github.com/user-attachments/assets/2fc4d865-af17-4af6-9dab-04543e4db7ae)

## Street - photo - inpaint (remove object)
Left: IP-Adapter | Right: "photo of a street in tokyo"

![benchmark_inpaint_street_input](https://github.com/user-attachments/assets/a9c770c6-677d-45d4-a558-4169d96547a6)

Xinsir Union ControlNet
![benchmark_inpaint_street](https://github.com/user-attachments/assets/9b6d2076-88b8-42db-8148-dcfad6b0e89c)

Diffusers Inpainting (Fooocus patch)
![benchmark_inpaint_street](https://github.com/user-attachments/assets/8f270d71-c9bb-4203-8c11-d4c49f870e88)

## Superman - photo - outpaint (expand)
Left: IP-Adapter | Right: "superman giving a speech at a congress hall filled with people"

![benchmark_inpaint_superman_input](https://github.com/user-attachments/assets/54bdd2d5-5f34-4c84-ba37-8abdb3636579)

Xinsir Union ControlNet
![benchmark_inpaint_superman](https://github.com/user-attachments/assets/037ad34d-f061-4671-b2e1-a5dcd301123c)

Diffusers Inpainting (Fooocus patch)
![benchmark_inpaint_superman](https://github.com/user-attachments/assets/e66c0df8-d8e4-4102-a5a1-d551786073f3)

## Torii - photo - inpaint (fill)
Left: IP-Adapter | Right: "photo of torii, japanese garden"

![benchmark_inpaint_tori_input](https://github.com/user-attachments/assets/b6939c03-fdf6-4caf-9ada-45768f5422b1)

Xinsir Union ControlNet
![benchmark_inpaint_tori](https://github.com/user-attachments/assets/49e7447a-5d61-4af8-85be-80ce1561e785)

Diffusers Inpainting (Fooocus patch)
![benchmark_inpaint_tori](https://github.com/user-attachments/assets/7fc99bb9-0887-460b-8b5d-c3032e7b7816)

## Conclusions

Some general trade-offs with the Xinsir ControlNet method:
* Less invasive and easier to tune than the Fooocus patch
* Model also works as ControlNet for scribble/lines/depth/normal/etc.
* Slower than the ad-hoc model merge by ~25%

Image quality is pretty good, but not convincingly better than previous method I'd say.

**Comment by BNP1111** [comment]
do u use V26model in fooocus or comfyui?

**Reply by Acly** [reply]
All tests run in ComfyUI, using identical workflow apart from the inpaint method. Native Fooocus pipeline has other differences that would impact results.

**Comment by minsky91** [comment]
An informative comparison, thanks! So it looks like Xinsir inpainting at present doesn’t qualify as replacement for the focus patch as the inpainting component in the Krita AI plugin, the seams and color aberrations are too evident in the outpainted areas.

I myself have already experimented a little with Xinsir under Forge, both for outpainting and inpainting (mostly for hi res images), the results are mixed. Actually, I couldn’t get it to work at all for anything that uses a mask, since, as you mentioned in your post in SD reddit, Xinsir promax takes as input the image with the masked area all black, I find it rather strange and unhelpful.

Don’t you know, there exists another inpaint model for SDXL, by Kataragi (https://huggingface.co/kataragi)? It’s quite capable, I liked the results of its outpainting, also under Forge - liked them way better than fooocus’s results with the same test image & prompts (the latter is also rather slow). Worth checking out, now that you are in the comparison mode :))

What I miss a lot in Krita AI diffusion plugin is the inpainting functionality that is available with the _inpaint_global_harmonious_ preprocessor under both A1111 and Forge (implementation in the latter is a bit incomplete though). It really delivers some magical results! I know there is no Comfy implementation of it, unfortunately (also according to mcmonkey of SwarmUI fame whom I asked about that on Discord), and it’s such a pity. Any chance to make a custom node for it in Comfy and bring it to the plugin eventually? If you haven’t seen what it does, I can show you examples. (Here’s one for starters:  https://civitai.com/articles/5834, a less-demanding SUPIR alternative).

**Reply by Acly** [reply]
> Xinsir promax takes as input the image with the masked area all black, I find it rather strange and unhelpful.

That's okay, all inpaint methods take an input like that indicating the mask, just some minor technical difference which made it incompatible with the SD1.5 inpaint pre-processor. In the end that's something the plugin (or preprocessor) does automatically anyway.

> Don’t you know, there exists another inpaint model for SDXL, by Kataragi

Apparently I bookmarked them and then forgot about it. Might as well run the tests with that model too.

> liked them way better than fooocus’s results with the same test image & prompts (the latter is also rather slow)

Why is it slower? The (fooocus) model merge should be the quickest since it has barely any overhead while sampling. In contrast, the ControlNet evaluation always adds 25~30%.

> inpaint_global_harmonious preprocessor

I never found any info on what that actually does, and reverse-engineering A1111 is super painful. I don't see how a pre-processor can do a lot of magic though.

**Comment by minsky91** [comment]
For an illustration, here's a nice quickie example of Kataragi's inpainting I just did in Forge, 1536x864 vector art frame of a fantasy landscape image using an enveloping mask. 

![image (52)](https://github.com/user-attachments/assets/4286914d-3a63-4965-b07d-5db5f58edc8d)
<img width="768" alt="Coddiwomple 06 - Leo AlbedoBaseXL Alchemy Environment + DarkArts 1536" src="https://github.com/user-attachments/assets/5378d642-2bd1-403e-9782-9a82f93df2fd">
![fractal landscape envelope mask](https://github.com/user-attachments/assets/6c384258-f1d1-4762-a6f0-36a80bbb26c5)

(most of my experiments are with outpainting actually, but the model works equally well with both, and it does support inpaint_global_harmonious)

**Comment by Acly** [comment]
Kataragi inpaint (weight=0.5, I also ran at 1.0 with worse results)

![benchmark_inpaint_apple-tree](https://github.com/user-attachments/assets/21064d85-6873-4b89-95bd-e20ff2f42509)
![benchmark_inpaint_bruges](https://github.com/user-attachments/assets/2ce6c35b-d3c3-4f12-bf6f-123a3817bf44)
![benchmark_inpaint_girl-cornfield](https://github.com/user-attachments/assets/9c0fd395-df19-46a4-88d2-c63d451807c5)
![benchmark_inpaint_jungle](https://github.com/user-attachments/assets/a937a58e-cfe4-44b1-aa37-2311ed5ac9c1)
![benchmark_inpaint_nature](https://github.com/user-attachments/assets/c55259b4-5f52-42bf-a36d-796abdcffd5f)
![benchmark_inpaint_park](https://github.com/user-attachments/assets/17a5f89f-d69c-4a29-b509-03e1b98446c1)
![benchmark_inpaint_street](https://github.com/user-attachments/assets/b814f53b-35a7-4be4-9396-79ec4ac939df)
![benchmark_inpaint_superman](https://github.com/user-attachments/assets/b90012ea-b40d-488c-98a6-39e7fb496a2f)
![benchmark_inpaint_tori](https://github.com/user-attachments/assets/9aea2cb2-b967-49db-8673-38f50564ecf2)

It's pretty good in some situations, but not a winner. Maybe it does better with artwork/anime since that seems to be its focus.

**Comment by minsky91** [comment]
It took me a while to investigate the various aspects of outpainting to address your comments, this is one of the more complicated areas of SD. Worse yet, outpainting via inpainting under Forge (which is my primary SD tool) works very differently, depending on the currently active tab: img2img, Inpaint or Inpaint Upload, even when using the same ControlNet settings, same input image/mask, same everything in short. Only the Inpaint tab can produce the artistic frame images which I used for the illustration above, they come out universally beautiful (with the right settings, of course). 

Timing. All SDXL inpainting models I tried worked fine for this task and rather fast: with the DPM++ 2M SDE Heun Exponential sampler and 24 steps they typically finished under 16 sec. after the initial preprocessing run (the card used was RTX 4070 Ti Super with 16 GB VRAM). Except for xinsir Promax, which took an unreasonably long time (8 min.) to process the same 1536x864 image and returned all black output. However, the settings in Inpaint must be chosen very carefully, for this to work. The main trick is to set the Ending Control Step very low - for some models, as low as the slider allows actually, i.e. 0.01. Testing outpainting was also difficult for the reason that the time it took for a model to process the image was inconsistent (even for the same input and parameters) and greatly varied from session to session, for no clear reason.

Before testing, I found two more inpaint ControlNet models on HuggingFace: [EcomXL Inpaint ControlNet](https://huggingface.co/alimama-creative/EcomXL_controlnet_inpaint) and Controlnet - [Inpainting dreamer by Desitech](https://huggingface.co/destitech/controlnet-inpaint-dreamer-sdxl), they work ok (ecomx slower than the rest though), but probably won’t be a competition to those we have already tried. 

inpaint26.foocus proved the slowest of the bunch: 30 sec. at ECS 0.25, the setting value at which it produced the best output. (This is despite what you might have read on the web about this particular aspect - [the suggestion is usually to set it to 0.5](https://github.com/lllyasviel/stable-diffusion-webui-forge/commit/fb2e2716683453e8b0ba775f7596d8c9b080177d)).  Going any lower than 0.25 would result in much slower initial processing times (up to 10 min.), with no visible benefits in the output.

Outpainting under fooocus, on the other hand, with the same input images and checkpoint and in Quality mode,  was very consistent, time-wise: always 30-32 sec., to produce 2456x1382 output (I don’t know why it always renders an upscaled version, there seems to be no way to disable it). Output quality with the Initial preset and default style, however, was nowhere near to what I am getting under Forge with the various inpainting models. Perhaps I should experiment with the Fooocus’s presets and styles, given the time.

I will return to inpaint_global_harmonious later, since I find it a very special preprocessor worthy of at least discussing, but it’s a bit of a separate topic.

**Reply by minsky91** [reply]
_Update_. This morning I have tested as well outpainting with Krita AI diff plugin (kind of logical, wouldn't it be?), and will post the results in a separate message, since they might - or should, even - influence further development of the plugin. Timings have been updated as well, and they look more favourable for the plugin, but it's not all.
